compiler/src/dotty/tools/dotc/typer/Synthesizer.scala

```diff
@@ -585,13 +595,34 @@ class Synthesizer(typer: Typer)(using @constructorOnly c: Context):
             tp.baseType(cls)
         val base = baseWithRefinements(formal)
         val result =
-          if (base <:< formal.widenExpr)
+          if (base <:< formal.widenExpr)  
             // With the subtype test we enforce that the searched type `formal` is of the right form
             handler(base, span)
-          else EmptyTree
-        result.orElse(recur(rest))
+          else EmptyTreeNoError
+        orElse(result, recur(rest))
       case Nil =>
-        EmptyTree
-    recur(specialHandlers)
+        EmptyTreeNoError
+    val result = recur(specialHandlers)
+    clearErrorsIfNotEmpty(result)
+
+end Synthesizer
+
+object Synthesizer:
+
+  /** Tuple used to store the synthesis result with a list of errors.  */ 
+  type TreeWithErrors = (Tree, List[String])
+  private def withNoErrors(tree: Tree): TreeWithErrors = (tree, List.empty)
+
+  private val EmptyTreeNoError: TreeWithErrors = withNoErrors(EmptyTree)
+
+  private def orElse(treeWithErrors1: TreeWithErrors, treeWithErrors2: => TreeWithErrors): TreeWithErrors = treeWithErrors1 match
+    case (tree, errors) if tree eq genericEmptyTree => 
+      val (tree2, errors2) = treeWithErrors2
+      (tree2, errors2 ::: errors)
```

Comment:
this puts the last errors at the front of the list, when I think each new error should go to the back
```suggestion
      (tree2, errors ::: errors2)
```